storage-operator/pkg/controller/volume/volume_controller.go

```diff
@@ -713,6 +714,38 @@ func (self *ReconcileVolume) finalizeVolume(
 	return delayedRequeue(nil)
 }
 
+func (self *ReconcileVolume) refreshDeviceName(
+	ctx context.Context,
+	volume *storagev1alpha1.Volume,
+	pv *corev1.PersistentVolume,
+) (reconcile.Result, error) {
+	path := pv.Spec.PersistentVolumeSource.Local.Path
+	nodeName := string(volume.Spec.NodeName)
+	reqLogger := log.WithValues(
+		"Volume.Name", volume.Name, "Volume.NodeName", nodeName,
+	)
+
+	name, err := self.salt.GetDeviceName(ctx, nodeName, volume.Name, path)
+	if err != nil {
+		self.recorder.Event(
+			volume, corev1.EventTypeNormal, "SaltCall",
+			"device path resolution failed",
+		)
+		reqLogger.Error(err, "cannot get device name from Salt response")
+		return delayedRequeue(err)
+	}
+	if pv.ObjectMeta.Annotations[DEVICE_ANNOTATION] != name {
+		metav1.SetMetaDataAnnotation(&pv.ObjectMeta, DEVICE_ANNOTATION, name)
+		if err := self.client.Update(ctx, pv); err != nil {
+			reqLogger.Error(err, "cannot update device name: requeue")
+			return delayedRequeue(err)
+		}
+		reqLogger.Info("update device name", "DeviceName", name)
+	}
+
+	return endReconciliation()
```

Comment:
This implies this call is made at the end of a reconciliation loop, right? Which is not necessarily the case?

In general, shouldn't only `Reconcile` ever return a `reconcile.Result`, and (e.g.) this one an `error`, and let `Reconcile` translate this in a `delayedRequeue` or whatever?

Reply:
> This implies this call is made at the end of a reconciliation loop, right? Which is not necessarily the case?

Not necessarily.
It may be ill-named, but the role of this function is to return a `reconcile.Result` that end the reconciliation (i.e. doesn't trigger a requeue), unlike `requeue` and `delayedRequeue`.

> In general, shouldn't only `Reconcile` ever return a `reconcile.Result`, and (e.g.) this one an `error`, and let `Reconcile` translate this in a `delayedRequeue` or whatever?

Hum, that would be like a global try/catch, where we inspect the "exception" and decide how to react to it.
Could be done, but would need to define a bunch of error types to be able to tell the difference between the different sources of error.
Could be done, but deserve a dedicated issue.

Reply:
> > This implies this call is made at the end of a reconciliation loop, right? Which is not necessarily the case?
> 
> Not necessarily.
> It may be ill-named, but the role of this function is to return a `reconcile.Result` that end the reconciliation (i.e. doesn't trigger a requeue), unlike `requeue` and `delayedRequeue`.

I know. However, imagine this were *not* the last step/call in `Reconcile`. Now in order to know whether to continue with the next operation or ask for a re-enqueue without running the 'next step', `Reconcile` would need to look 'inside' this `reconcile.Result`...

> 
> > In general, shouldn't only `Reconcile` ever return a `reconcile.Result`, and (e.g.) this one an `error`, and let `Reconcile` translate this in a `delayedRequeue` or whatever?
> 
> Hum, that would be like a global try/catch, where we inspect the "exception" and decide how to react to it.
> Could be done, but would need to define a bunch of error types to be able to tell the difference between the different sources of error.
> Could be done, but deserve a dedicated issue.

Oh, for sure, out of scope.

Maybe the API could be as follows: instead of `(reconcile.Result, error)`, the functions could *still* return `(reconcile.Result, error)`, but the `Result` *must* be `nil` when `error` is `nil`, and may *not* be `nil` when `error` is not `nil`. As such, the `Result` is returned after the `error` is logged on error, otherwise `Reconcile` continues whatever it needs to do next, and in the end returns its own 'success' `Result`.

Reply:
Would be better yeah, I'll open a debt ticket for that.